distribution/lib/Standard/Base/0.0.0-dev/src/System/File/Generic/Writable_File.enso

```diff
@@ -0,0 +1,84 @@
+import project.System.File.Generic.File_Like.File_Like
+import project.System.File.Existing_File_Behavior.Existing_File_Behavior
+import project.System.File.File
+import project.System.File.File_Access.File_Access
+import project.System.File_Format_Metadata.File_Format_Metadata
+import project.System.Output_Stream.Output_Stream
+from project.Data.Boolean import Boolean, False, True
+
+## PRIVATE
+   Any type that represents a file that can be written to should be convertible
+   to this common type-class.
+
+   This allows our code to work with any kind of file backend, e.g. the local
+   file system as well as remote ones like S3, using the same basic interface.
+
+   The type-class currently provides the following capabilities:
+   - write_strategy: a `File_Write_Strategy` instance denoting how to handle
+     file writing depending on `Existing_File_Behavior` and dry run settings.
+   - exists: the underlying `file` should have an `exists` method available,
+     because it is used by some logic as an early exit condition.
+type Writable_File
+    ## PRIVATE
+    Value file (write_strategy : File_Write_Strategy)
+
+    ## PRIVATE
+    exists self -> Boolean = self.file.exists
+
+    ## PRIVATE
+    copy_to self destination replace_existing = self.file.copy_to destination replace_existing
+
+    ## PRIVATE
+    write self (existing_file_behavior : Existing_File_Behavior) (action : Output_Stream -> Any) -> Any =
+        self.write_strategy.write self.file existing_file_behavior action
+
+    ## PRIVATE
+    write_handling_dry_run self (existing_file_behavior : Existing_File_Behavior) (action : Any -> Output_Stream -> Any) -> Any =
+        self.write_strategy.write_handling_dry_run self.file existing_file_behavior action
+
+    ## PRIVATE
+    is_local self -> Boolean = self.file.is_a File
+
+    ## PRIVATE
+       This helper method may be used if the logic for writing the file requires a local filesystem file
+       (e.g. using some external libraries that prefer to work with a Java File).
+       It downloads a remote file to a local temporary location, runs the action
+       and then uploads the resulting file back to the remote backend.
+    write_requiring_local_file self (existing_file_behavior : Existing_File_Behavior) (action : File -> Any) -> Any =
+        self.write_strategy.write_with_local_file self.file existing_file_behavior action
+
+    ## PRIVATE
+    with_output_stream self (open_options : Vector) action =
+        self.file.with_output_stream open_options action
+
+    ## PRIVATE
+    to_text self -> Text =
+        "(Writable_File " + self.file.to_text + ")"
+
+    ## PRIVATE
+    pretty self -> Text = self.to_text
+
+    ## PRIVATE
+    to_display_text self -> Text = self.file.to_display_text
+
+## PRIVATE
+Writable_File.from (that : Text) =
+    ## Currently this only works for local filesystem paths
+       TODO We should extend it to also support custom paths like S3, through a ServiceProvider solution
```

Comment:
Isn't this already covered by `File.new` which I thought I had put hooks into for exactly that?
Feels like it should be covered in the single place.

Reply:
Yours is for `File.from Text` this one is for `Writable_File.from Text`, so they are a bit different.

I think overall we will need to go towards `Writable_File.from Text` and `File_Like.from Text` to support these paths.

Just `File.from Text` won't work with SPI as it has to return `File` type but that may return other types like `S3_File` or others.

But yeah I think we will need to unify this, I imagine we could have something like:
```
File_Like.from (that : Text) = resolve using SPI
```

and then the `File_Like` will allow to convert itself to other 'classes' like `Writable_File`.

I think we should clean this up as part of the ticket implementing this SPI.